sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategy.scala

```diff
@@ -272,7 +272,10 @@ object FileSourceStrategy extends Strategy with PredicateHelper with Logging {
             case FileFormat.ROW_INDEX =>
               fileFormatReaderGeneratedMetadataColumns
                 .find(_.name == FileFormat.ROW_INDEX_TEMPORARY_COLUMN_NAME)
-                .get.withName(FileFormat.ROW_INDEX)
+                // Change the `_tmp_metadata_row_index` to `row_index`,
+                // and also change the nullability to not nullable,
+                // which is consistent with the nullability of `row_index` field
+                .get.withName(FileFormat.ROW_INDEX).withNullability(false)
```

Comment:
shall we update `fileFormatReaderGeneratedMetadataColumns` to set nullablity as false?

Reply:
Thanks, Wenchen, I tried that before, but it failed many test cases in `FileMetadataStructRowIndexSuite`. See [this fix commit](https://github.com/apache/spark/pull/38777/commits/0f98dd97a3c84a10be6fd7f75468e5d62f17fe82) and [this comment](https://github.com/apache/spark/pull/38777#issuecomment-1332079823).

I don't have much context on the `row_index`, not sure what caused the issue, any idea? Thanks!